src/kcas.ml

```diff
@@ -194,35 +238,43 @@ module Op = struct
   type t = CAS : 'a Loc.t * 'a * 'a -> t
 
   let make_cas loc before after = CAS (loc, before, after) [@@inline]
+  let make_cmp loc expected = CAS (loc, expected, expected) [@@inline]
 
   let is_on_loc op loc =
     match op with CAS (loc', _, _) -> Obj.magic loc' == loc
     [@@inline]
 
   let atomic = function
-    | CAS (loc, before, after) -> Loc.compare_and_set loc before after
+    | CAS (loc, before, after) ->
+        if before == after then Loc.get loc == before
+        else Loc.compare_and_set loc before after
 
-  let atomically = function
+  let atomically ?(mode = Mode.lock_free) = function
     | [] -> true
     | [ op ] -> atomic op
     | first :: rest ->
-        let casn = Atomic.make `After in
-        let insert cass = function
-          | CAS (loc, before, after) -> insert cass loc { before; after; casn }
+        let casn = Atomic.make (mode :> status) in
+        let rec run cass = function
+          | [] -> determine_for_owner casn cass
+          | CAS (loc, before, after) :: rest ->
+              if before == after && is_obstruction_free casn then
```

Comment:
Mode is given as argument but it's re-read from atomic var here. Can mode actually change mid-way through execution?

Reply:
No, it cannot change.

The `casn` atomic location is otherwise unused until the full `cass` splay tree is ready and then the `casn` is set to `Undetermined cass` in `determine_for_owner` before calling `determine`.

I pass the mode through the `casn` in the transaction mechanism, because that way every `Tx` operation will potentially allocate one word less:

```diff
-type log = casn * cass
+type log = mode * casn * cass
```

I then use the same piggybacking in `atomically` for consistency.

Admittedly it is a bit of hack/might not be immediately obvious, but I tried a number of approaches and this seemed to give low overhead and `is_obstruction_free casn` accurately reads as "is this CASN going to be attempted in obstruction free mode".